src/de/serde.rs

```diff
@@ -212,13 +218,191 @@ impl<'de> Visitor<'de> for BsonVisitor {
         Ok(Bson::Array(values))
     }
 
-    #[inline]
-    fn visit_map<V>(self, visitor: V) -> Result<Bson, V::Error>
+    fn visit_map<V>(self, mut visitor: V) -> Result<Bson, V::Error>
     where
         V: MapAccess<'de>,
     {
-        let values = DocumentVisitor::new().visit_map(visitor)?;
-        Ok(Bson::from_extended_document(values))
```

Comment:
The logic here is much the same, but it was updated to make use of serde's `MapAccess` API for the extended JSON format parts to achieve better performance. Prior to these changes, a `Document` would allocated and deserialized each time, and then its format would be parsed. Instead, we deserialize the extended format directly, allowing us to go straight from Raw BSON to our BSON types where possible. If the input data doesn't match an extended JSON format, it's just deserialized to a `Document`.